### PR TITLE
subsys: bootloader: fix breaking changes introduced by kconfig after DTC

### PR DESCRIPTION
**Pull request by sigvartmh**
Fixes secure bootloader so that it can be compiled with master for nrf52840 and nrf9160 due to the breaking changes when moving Kconfig after DTC.

### CODE REVIEW

**Comment by sigvartmh** [comment]
@hakonfam @oyvindronningstad @osaether

**Comment by sigvartmh** [comment]
@hakonfam @oyvindronningstad @osaether

**Comment by hakonfam** [comment]
@carlescufi  This i a blocker for all other bootloader related PRs, so a rapid response would be greatly appreciated :)

File: subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c [hunk 1]
```diff
@@ -26,8 +26,14 @@ void __secure_fault(void) ALIAS("dummy_handler");
 void __svc(void) ALIAS("dummy_handler");
 void __pendsv(void) ALIAS("dummy_handler");
 void __sys_tick(void) ALIAS("dummy_handler");
+void z_clock_isr(void) ALIAS("dummy_handler");
```
Comment (hakonfam) [inline subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c]: Could you put this in a separate commit?

**Comment by hakonfam** [comment]
The commit "subsys: bootloader: cmake: add overide to provision script"  should have different tags as it only changes a python file. "scripts: bootloader:" perhaps.

File: scripts/bootloader/provision.py [hunk 2]
```diff
@@ -55,6 +55,9 @@ def parse_args():
         formatter_class=argparse.RawDescriptionHelpFormatter)
 
     parser.add_argument("--generated-conf-file", required=True, help="Generated conf file.")
+    parser.add_argument("--s0-addr", type=int, required=False, help="Set address of s0 explicit")
```
Comment (oyvindronningstad) [inline scripts/bootloader/provision.py]: In next three lines:
explicit -> explicitly

File: scripts/bootloader/provision.py [hunk 3]
```diff
@@ -55,6 +55,9 @@ def parse_args():
...
+    parser.add_argument("--s1-addr", type=int, required=False, help="Set address of s1 explicit")
+    parser.add_argument("--provision-addr", type=int, required=False, help="Set address of proision data explicit")
```
Comment (oyvindronningstad) [inline scripts/bootloader/provision.py]: proision -> provision

File: subsys/bootloader/CMakeLists.txt [hunk 4]
```diff
@@ -177,13 +169,32 @@ add_custom_target(
   offsets_h
   )
 
+if(CONFIG_SOC_NRF52840)
+	set(OVERRIDE
```
Comment (oyvindronningstad) [inline subsys/bootloader/CMakeLists.txt]: I suggest
OVERRIDE -> ADDR_OVERRIDE

File: subsys/bootloader/bootloader.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
+/*TODO: Remove when multi-image has been merged*/
```
Comment (oyvindronningstad) [inline subsys/bootloader/bootloader.h]: Why not put the addresses in Kconfig? That would make them available in CMake as well.

File: subsys/bootloader/c_runtime_setup/Kconfig [hunk 6]
```diff
@@ -4,17 +4,9 @@
...
-		bool "Custom C runtime setup"
-	config SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
-		bool "Zephyr C runtime setup"
-	config SB_C_RUNTIME_SETUP_VARIANT_MDK
-		bool "MDK C runtime setup"
-		select SB_VENDOR_SYSTEM_INIT
-endchoice
+config SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
+	bool "Zephyr C runtime setup"
+	default y
```
Comment (oyvindronningstad) [inline subsys/bootloader/c_runtime_setup/Kconfig]: What happens if this is set to 'n'?

**Comment by oyvindronningstad** [comment]
The commit "subsys: bootloader: Remove default Kconfig values" could have a more descriptive commit message:
 - Which default Kconfig values?
 - Why can these choices no longer be used?

File: subsys/bootloader/bootloader.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+#ifndef BOOTLOADER_H__
+#define BOOTLOADER_H__
+
+#if CONFIG_SOC_NRF52840
+/* partition@0 */
+#define DT_FLASH_AREA_SECURE_BOOT_OFFSET	0
+#define DT_FLASH_AREA_SECURE_BOOT_SIZE	16384
```
[the remark this answers is not in the document]
Comment (hakonfam) [reply subsys/bootloader/bootloader.h]: Note that all these are replaced when we get multi image support in the build system.

[the remark this answers is not in the document]
Comment (hakonfam) [reply subsys/bootloader/bootloader.h]: But I agree :)

File: subsys/bootloader/c_runtime_setup/Kconfig (hunk 6 above)
Comment (sigvartmh) [reply subsys/bootloader/c_runtime_setup/Kconfig]: Compile error, I'll just remove the prompt

**Comment by sigvartmh** [comment]
> The commit "subsys: bootloader: Remove default Kconfig values" could have a more descriptive commit message:
> 
>     * Which default Kconfig values?
> 
>     * Why can these choices no longer be used?

I'll add some more text explaining the rational

File: subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c (hunk 1 above)
Comment (sigvartmh) [reply subsys/bootloader/c_runtime_setup/zephyr_dummy_interrupt_handlers.c]: Sure

File: subsys/bootloader/bootloader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: It's DTC output but sure I can try to format them back

File: subsys/bootloader/bootloader.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+/* partition@10000 */
+#define DT_FLASH_AREA_S1_OFFSET	65536
+#define DT_FLASH_AREA_S1_SIZE		49152
+
+/* partition@1C000 */
+#define DT_FLASH_AREA_APP_OFFSET	114688
+#define DT_FLASH_AREA_APP_SIZE	929792
+
+/* partition@4000 */
+#define DT_FLASH_AREA_S0_OFFSET		16384
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: Type First->Second->Third->ETC... ?

File: subsys/bootloader/bootloader.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+#define DT_FLASH_AREA_S0_SIZE		49152
+
+/* partition@4000 */
+#define DT_FLASH_AREA_S0_OFFSET	16384
+#define DT_FLASH_AREA_S0_SIZE		49152
+
+/* partition@FF000 */
+#define DT_FLASH_AREA_PROVISION_OFFSET	1044480
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: You could do `FLASH_SIZE-PROVISION_SIZE`

File: subsys/bootloader/bootloader.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+/* partition@10000 */
+#define DT_FLASH_AREA_APP_OFFSET	65536
+#define DT_FLASH_AREA_APP_SIZE	978944
+
+/* partition@8000 */
+#define DT_FLASH_AREA_S0_OFFSET	32768
+#define DT_FLASH_AREA_S0_SIZE		16384
+
+/* partition@C0000 */
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: DTC output so did not really check if comments are correct

File: subsys/bootloader/bootloader.h (hunk 5 above)
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: The only reasoning was less diff from now to multi image. Since the variable names will be prefixed with CONFIG_ instead of DT_. 

This is a fixup until multi-image is merged when multi image is merged removing `bootloader.h` would give you compile errors so it's easier to remove where it's used. Instead of modifying the Kconfig file then revert. All files touched but it's open for discussion if you would rather prefer it placed in Kconfig.

File: subsys/bootloader/bootloader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/bootloader/bootloader.h]: Oh, I see, you just took the generated file. In that case it's not a big deal, since it is temporary.

File: subsys/bootloader/bootloader.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/bootloader/bootloader.h]: Just the order they appear in flash.
secure boot -> s0 -> s1 -> app -> provision.
Again, if you just took generated output, it's not a big deal.

File: subsys/bootloader/bootloader.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/bootloader/bootloader.h]: Yes, that's what I was thinking. That should be sufficient to support all nRF5x devices.

File: subsys/bootloader/bootloader.h (hunk 5 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bootloader.h]: Ok, I was wondering if we could reuse the values for the provision override parameters in the cmake file. It's maybe not worth it at this point, at least if it will change again.

Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: I don't think it's worth it at this point, as we would have to change it again. So it's better to change it when we have the actual build system in place :1st_place_medal:

File: subsys/bootloader/bootloader.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: Yeah it works for the nrf5 devices. Couldn't find a general config variable for it so it will have to be
```
#if CONFIG_SOC_NRF52840  ||\
    CONFIG_SOC_NRF52832  ||\
    CONFIG_SOC_NRF52810  ||\
    CONFIG_NRF51822_QFAA ||\
    CONFIG_NRF51822_QFAB ||\
    CONFIG_NRF51822_QFAC
```

File: scripts/bootloader/provision.py (hunk 2 above)
Comment (sigvartmh) [reply scripts/bootloader/provision.py]: fixed (not pushed)

File: scripts/bootloader/provision.py (hunk 3 above)
Comment (sigvartmh) [reply scripts/bootloader/provision.py]: fixed(not pushed)

File: subsys/bootloader/CMakeLists.txt (hunk 4 above)
Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: Renamed(not pushed)

File: subsys/bootloader/bootloader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: Fixed(not pushed)

File: subsys/bootloader/bootloader.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: fixed(not pushed)

File: subsys/bootloader/bootloader.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: Done

File: subsys/bootloader/bootloader.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/bootloader.h]: Done

File: subsys/bootloader/c_runtime_setup/Kconfig (hunk 6 above)
Comment (sigvartmh) [reply subsys/bootloader/c_runtime_setup/Kconfig]: Done

**Comment by sigvartmh** [comment]
This PR has been **updated**. Most notable changes are done in `bootloader.h` other than that the commit history has been rewritten and split up into logically preceding commits.
*  Kconfig symbol for changing linkerscript/startup code is deprecated and removed
* More unused files are removed hence the ~-4000LOC

File: subsys/bootloader/bootloader.c [hunk 11]
```diff
@@ -86,9 +86,6 @@ void uninit_used_peripherals(void)
 }
 
 #ifdef CONFIG_SW_VECTOR_RELAY
-#ifndef CONFIG_SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
-_GENERIC_SECTION(.vt_pointer_section) u32_t _vector_table_pointer;
```
Comment (oyvindronningstad) [inline subsys/bootloader/bootloader.c]: Are you sure this line is not needed?
It's for nRF51 which has no VTOR.

Comment (sigvartmh) [reply subsys/bootloader/bootloader.c]: I could be wrong but I thought it was not needed if `SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR` is standard and that it defines this `extern  _vector_table_pointer` in some other file am I wrong?

**Comment by sigvartmh** [comment]
@nvlsianpu This is ready to be reviewed

File: subsys/bootloader/bootloader.c (hunk 11 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bootloader.c]: Oh, sorry I misread, it's #ifndef not #ifdef.

File: subsys/bootloader/CMakeLists.txt [hunk 12]
```diff
@@ -177,13 +169,32 @@ add_custom_target(
   offsets_h
   )
 
+if(CONFIG_SOC_NRF52840)
+	set(ADDR_OVERRIDE
+		--s0-addr 16384
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: Oh must have overlooked it will change it to hex

File: subsys/bootloader/include/linker.ld [hunk 13]
```diff
@@ -111,6 +132,17 @@ SECTIONS
 	KEEP(*(.dbghdr))
 	KEEP(*(".dbghdr.*"))
 #endif
+
+#ifdef CONFIG_NXP_IMX_RT_BOOT_HEADER
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/include/linker.ld]: Yes, this file is temporary until multi-image is merged.

File: subsys/bootloader/c_runtime_setup/CMakeLists.txt [hunk 14]
```diff
@@ -52,68 +40,8 @@ target_include_directories_ifdef(CONFIG_SB_C_RUNTIME_SETUP_VARIANT_ZEPHYR
...
-  PRIVATE
-  ${HAL}/nordic/nrfx/mdk/system_nrf9160.c
-  )
-target_sources_ifdef(CONFIG_SB_C_RUNTIME_SETUP_VARIANT_MDK
-  c_runtime_setup
-  PRIVATE
-  gcc_startup_nrf9160.S
-  )
-endif()
+add_library(kernel_init STATIC ${ZEPHYR_BASE}/kernel/init.c)
```
Comment (SebastianBoe) [inline subsys/bootloader/c_runtime_setup/CMakeLists.txt]: This isn't a zephyr library, because it should be linked into the bootloader, and not the app, right?

Please add a comment that explains this.

Then I wouldn't be confused again, and it will be more clear that this needs to change after the multi-image migration.

File: subsys/bootloader/CMakeLists.txt [hunk 15]
```diff
@@ -177,13 +169,32 @@ add_custom_target(
...
+		--s1-addr 65536
+		--provision-addr 520192)
+elseif(CONFIG_SOC_NRF9160)
+	set(ADDR_OVERRIDE
+		--s0-addr 32768
+		--s1-addr 786432
+		--provision-addr 1044480)
+endif()
+
+
```
Comment (SebastianBoe) [inline subsys/bootloader/CMakeLists.txt]: DT_FLASH_AREA_S0_OFFSET is now available in the DT namespace (you can do ${DT_FLASH_AREA_S0_OFFSET}).

So I think you want to remove the generated_dts_board.conf parser function, and instead determine the possibly overridden value in CMake. instead of in python. Ask me for details. I could be wrong.

File: subsys/bootloader/c_runtime_setup/CMakeLists.txt (hunk 14 above)
Comment (sigvartmh) [reply subsys/bootloader/c_runtime_setup/CMakeLists.txt]: Yes, you are correct it isn't a zephyr library it was a hackish way of being able to set `_Cstart=UNUSED_Cstart` only for compiling the C runtime/startup code of zephyr then override the _Cstart rutine with our bootloaders `main_bl` call. I'll add a comment explaining this. Do you mind elaborating a bit on what needs to be changed after multi-image ?

File: subsys/bootloader/CMakeLists.txt (hunk 15 above)
Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: It's a great idea unfortuntaly these symbols does not yet exsists in the DT https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/1d6219ffec845a43f66469354d997fbea2f283a4/boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts#L165

I'll investigate more into your 2nd comment I'll ask you tomorrow :)

File: subsys/bootloader/c_runtime_setup/CMakeLists.txt (hunk 14 above)
Comment (sigvartmh) [reply subsys/bootloader/c_runtime_setup/CMakeLists.txt]: Added comment @SebastianBoe check if the comment gives enough information so that you are not confused by the purpose of this lib.

File: subsys/bootloader/CMakeLists.txt (hunk 12 above)
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: Fixed

File: subsys/bootloader/CMakeLists.txt (hunk 15 above)
Comment (SebastianBoe) [reply subsys/bootloader/CMakeLists.txt]: I can't read DT source code and predict what symbols will be available ;)

The reason I believed they were available is that when I build, I get symbols like these in CMake:

```
DT_FLASH_AREA_MCUBOOT_OFFSET=0
DT_FLASH_AREA_STORAGE_OFFSET=122880

```

But I haven't tried in a build that has the S0 area defined.

File: subsys/bootloader/c_runtime_setup/CMakeLists.txt (hunk 14 above)
Comment (SebastianBoe) [reply subsys/bootloader/c_runtime_setup/CMakeLists.txt]: LGTM

File: subsys/bootloader/CMakeLists.txt (hunk 15 above)
Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: Trying to predict naming and which symbols are available from DT is quite hard and it was a nice guess the only thing missing to get the symbols you are asking for is 

```
storage_partition: partition@(offset) { //@has no effect and is just inferred as a comment)
			label = "s0";
			reg = <(offset) (size)>;
};
```
in the partition section but this will have to be added for every board instead of every SoC I think? Honestly I'm not sure what would be best, this would at least give the smallest diff towards zephyr if we don't want to try to upstream partitioning changes upstream.

A problem with trying to upstream it is that MCUBoot is relaying on the mcuboot symbols as they can't overlay their own DT overlay file outside the zephyr project folder as overlays have to be board specific to be overlayed

Comment (sigvartmh) [reply subsys/bootloader/CMakeLists.txt]: Moved the code out of python and into cmake now so that instead of an override we do an fallback if the values does not exsist in DT

Comment (SebastianBoe) [reply subsys/bootloader/CMakeLists.txt]: Great!
